### ISSUE TITLE
File Provider issues in Razor Pages apps

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/aspnetcore#8971]
### Describe the bug
I created a Razor Pages app with vb.net. I use a virtual File Provider to deliver the page which I am crating using xml literals in vb code.
I am facing some issues regarding body pages. Take the Index.cshtml page as an example:
1. if you deleted the Index.cshtml and run, the browser shows a blank white page! Why there is no 404 page not found message?

2. If the file Index.cshtml exists in the Pages folder, the routing system use it directly and doesn't send a request to the file provider, which is not the expected behavior: ask the provider first, then look fot the physical path.
OK, I deleted the cshtml file, so the file provider received a request for it. But my problem is that the IndexModel.OnGet event is not fired because the IndexModel class is registered in the cshtml which doesn't exist yet!
 I want an event like OnGet fired after the browser sent the request to use it to prepare my virtual page and send it to the file provider.
The only workaround I can try know, is to fire an event from my Implementation of the file provider, and listen to it in all model classes and the one that corresponds to the requested file do the work. Obviously this will have a bad impact on the server to call tens of event handlers for every single request!
Is there any solution for this? What is the event that fires directly after the route is requested and before trying to locate the cshtml file?

3. The file "/_ViewImports.cshtml" is requested 9 times from the file provider from running the app until the Index page is viewed in the browser! And each page requested 2 times! why? Can this be optimized?

### To Reproduce
You can use the Vazor.sln  solution from this repo:
https://github.com/VBAndCs/Vazor 
The solution contains VazorPages1 project and the Vazor project which contains the VazorViewProvider, the file provider.
 
# Note:
In the VazorPages1 I left the Index.cshtml file as an entry point, and injected my vbxml code into it as a partial view like this:
```
 @Html.Partial(Model.ViewName + ".cshtml")
```

the Model.ViewName holds a unique name for the partial view for each request, because these are evaluated pages that can be used for its request only. The partial view is delivered by the file provider.
So, VazorPages1 is working fine by this twisted workaround!

**Comment by VBAndCs** [comment]
I updated the issue to add a third note:
> The file "/_ViewImports.cshtml" is requested 9 times from the file provider from running the app until the Index page is viewed in the browser! And each page requested 2 times! why? Can this be optimized?

**Comment by VBAndCs** [comment]
I used a workaround to virtually generate the entry cshtml like this:
```VB.NET
    Private Shared Function CreatePage(viewName As String) As String
        Return _
$"@page
@model {viewName}Model

    <div>
        @Html.Partial(Model.ViewName + {""""}.cshtml{""""})
    </div>"
    End Function
```

It works fine, but my 3 notes are still valid. Ther eis a new note about ASP.NET Core 2.2:
The routing system doesn't use the File provider even it didn't find the physical .cshtml file, but unlike ASP.NET Core 3, the browser shows a "page not found" page!

I still need a way to intercept the route in Razor Pages app before the router tries to locate the page, because as I mensioned, the PageModel.OnGet doesn't exist before the page is located.
Thanks.

**Comment by VBAndCs** [comment]
@Eilon @danroth27
I faced a new issue with file provider (using .net core 2.2):
The app contains Pages\Index.cshtml and Pages\Basket\Index.cshtml
Both contain `@model IndexModel` directive.
Without using file provider, navigating to both pages works fine, but when using file provider with both pages still physically in same folders (i.e file provider doesn't do anything yet), navigating to the Pages\Basket\Index.cshtml pages invokes the `Pages\ Index.cshtml.vb` class instead of the `Pages\Basket\Index.cshtml.vb` class!
I changed the @model directive in the `Pages\Basket\Index.cshtml` page to be `@model Pages.Basket.IndexModel` and everything worked correctly!
The Pages/Basket folder doesn't contain a `_ViewImports.cshtml` page, but it works correctly without file provider.
Why this difference?
